### PR TITLE
Autoselect location if available

### PR DESCRIPTION
**Pull request by esoergel**
http://manage.dimagi.com/default.asp?170383
@snopoke
I've verified that with this change, you can access your supply_point_id with `"instance('commcaresession')/session/data/supply_point_id"` if your user has one.  If your user does not have an assigned location, the app fails with the following message:

> Filtering your list failed with a bad filter expression.  Please fix this and re-deploy your app.  Error was XPath evaluation: type mismatch Location instance(commcaresession/session/user/data/commtrack-supply-point was not found.

### CODE REVIEW

File: corehq/apps/app_manager/suite_xml.py [hunk 1]
```diff
@@ -1691,6 +1707,12 @@ def get_datum_meta_module(self, module, use_filter=False):
             })
         return datums
 
+    def get_userdata_autoselect(self, key, session_id, mode):
+        xpath = session_var(key, path='user/data')
+        datum = SessionDatum(id=session_id, function=xpath)
+        assertions = self.get_auto_select_assertions(xpath, mode, [key])
+        return datum, assertions
```
Comment (esoergel) [inline corehq/apps/app_manager/suite_xml.py]: I moved this from higher up, 'cause the only place this is used now is just below.  Think I should un-abstract it now that it's a one-off?

File: corehq/apps/app_manager/tests/data/suite/autoload_supplypoint.xml [hunk 2]
```diff
@@ -14,14 +14,7 @@
...
     <assertions>
-      <assert test="count(instance('commcaresession')/session/user/data/commtrack-supply-point) = 1">
-        <text>
-          <locale id="case_autoload.location.property_missing">
-            <argument>commtrack-supply-point</argument>
-          </locale>
-        </text>
-      </assert>
-      <assert test="count(instance('casedb')/casedb/case[@case_id=instance('commcaresession')/session/user/data/commtrack-supply-point]) = 1">
+      <assert test="if(count(instance('commcaresession')/session/user/data/commtrack-supply-point) = 1, count(instance('casedb')/casedb/case[@case_id=instance('commcaresession')/session/user/data/commtrack-supply-point]) = 1, true())">
```
Comment (esoergel) [inline corehq/apps/app_manager/tests/data/suite/autoload_supplypoint.xml]: unwrapped, this looks like:
```
if(count(instance('commcaresession')/session/user/data/commtrack-supply-point) = 1,
   count(instance('casedb')/casedb/case[
           @case_id=instance('commcaresession')/session/user/data/commtrack-supply-point
         ]) = 1,
   true())
```

Comment (snopoke) [inline corehq/apps/app_manager/tests/data/suite/autoload_supplypoint.xml]: Will this still not raise an error loading the datum if there is no `commtrack-supply-point` element?

I think the function expression for the datum may also have to be an if.

Also, if the user has no user data at all then the `instance('commcaresession')/session/user/data` won't exist either which is what caused the original bug.

File: corehq/apps/app_manager/tests/data/suite/suite-advanced-commtrack.xml [hunk 3]
```diff
@@ -373,7 +408,15 @@
     <session>
       <datum id="case_id_case_requisition" nodeset="instance('casedb')/casedb/case[@case_type='requisition'][@status='open']" value="./@case_id" detail-select="m2_case_short" detail-confirm="m2_case_long"/>
       <datum id="product_id" nodeset="instance('commtrack:products')/products/product[program_id='abc123def']" value="./@id" detail-select="m2_product_short"/>
+      <datum id="supply_point_id" function="instance('commcaresession')/session/user/data/commtrack-supply-point"/>
```
Comment (esoergel) [inline corehq/apps/app_manager/tests/data/suite/suite-advanced-commtrack.xml]: make this fall back to "not-a-real-supply-point" or similar

File: corehq/apps/app_manager/suite_xml.py [hunk 4]
```diff
@@ -1449,6 +1455,16 @@ def entries(self):
...
 
+                if self.app.commtrack_enabled:
+                    from .models import AUTO_SELECT_LOCATION
+                    datum, assertions = self.get_location_assertions(
+                        'commtrack-supply-point',
+                        'supply_point_id',
+                        AUTO_SELECT_LOCATION,
+                    )
+                    e.datums.append(datum)
+                    e.assertions.extend(assertions)
```
Comment (esoergel) [inline corehq/apps/app_manager/suite_xml.py]: See if `"instance('commcaresession')/session/data/supply_point_id" in form.source()`

**Comment by esoergel** [comment]
@snopoke I just tried this out on staging.  If your user does not have an assigned location, the app fails with the following message when you try to open the form:

> Filtering your list failed with a bad filter expression. Please fix this and re-deploy your app. Error was XPath evaluation: type mismatch Location instance(commcaresession/session/user/data/commtrack-supply-point was not found.

That looks like a mobile message, any idea why it's not showing the proper error?

**Comment by snopoke** [comment]
Can you point me to the app on staging and also the user that you're testing with

**Comment by esoergel** [comment]
@snopoke It should be g2g now.  It's on staging.  I verified that it works as expected with users which have locations, and for those that don't, the app fails when they attempt to enter the form.  The error message is
> The location specified for case auto-selecting could not be found: ${0}commtrack-supply-point

Any idea why that extra `${0}` is there?  This uses the same help text builder as the other case autoselect stuff.  I think we should maybe special-case the error message here, since we have more assumptions about what's happening here than we do elsewhere.  Something like 
> This form requires access to the user's location, but none was found.
and if the case is missing:
> This form requires the user's location to have a supply-point case.

Either way, I think this is good enough to merge, then we can clean up the error message separately.

**Comment by snopoke** [comment]
+1 on the change in language. Not sure why the ${0} is showing up

**Comment by snopoke** [comment]
Nice, looks good. Merge once tests pass.

**Pull request merged**